main.c

```diff
@@ -36,98 +36,54 @@ struct {
 
 void main_initthr(void *unused)
 {
-	size_t i;
-	syspage_program_t prog;
-	int xcount = 0, res;
-	char *cmdline = syspage->arg, *end;
-	char *argv[32], *arg, *argend;
+	int res;
+	unsigned int argc;
+
+	syspage_prog_t *prog;
+	char *argv[32], *cmdline;
 
 	/* Enable locking and multithreading related mechanisms */
 	_hal_start();
 
-	lib_printf("main: Decoding programs from data segment\n");
-	programs_decode(&main_common.kmap, &main_common.kernel);
-
 	lib_printf("main: Starting syspage programs:");
-	for (i = 0; i < syspage->progssz; i++)
-		lib_printf(" '%s',", syspage->progs[i].cmdline);
-	lib_printf("\b \n");
+	syspage_progShow();
 
 	posix_init();
 	posix_clone(-1);
 
-	/* Free memory used by initial stack */
-	/*vm_munmap(&main_common.kmap, main_common.stack, main_common.stacksz);
-	vm_pageFree(p);*/
-
-	/* Set stdin, stdout, stderr ports */
-//	proc_fileAdd(&h, &oid, 0);
-//	proc_fileAdd(&h, &oid, 0);
-//	proc_fileAdd(&h, &oid, 0);
-
-	argv[0] = NULL;
-
-	while (cmdline != NULL && *cmdline != '\0') {
-		end = cmdline;
-		while (*end && *(++end) != ' ');
-		while (*end == ' ')
-			*(end++) = 0;
-		if (*cmdline == 'X' && ++xcount) {
-			i = 0;
-			argend = cmdline;
-
-			while (i < sizeof(argv) / sizeof(*argv) - 1) {
-				arg = ++argend;
-				while (*argend && *argend != ';')
-					argend++;
+	/* Start programs from syspage */
+	if ((prog = syspage_progList()) != NULL) {
+		do {
+			cmdline = prog->argv;
+			/* If app shouldn't be executed then args should be discarded */
+			if (*cmdline != 'X') {
+				while (*cmdline != ';' && *cmdline != '\0')
+					++cmdline;
+
+				*cmdline = '\0';
+				continue;
+			}
 
-				argv[i++] = arg;
+			/* 'X' is no longer useful */
+			++prog->argv;
+			cmdline = prog->argv;
+			for (argc = 0; argc < (sizeof(argv) / sizeof(*argv) - 1);) {
+				argv[argc++] = cmdline;
+				while (*cmdline != ';' && *cmdline != '\0')
+					++cmdline;
```

Comment:
```suggestion
			for (argc = 0; (argc < (sizeof(argv) / sizeof(*argv) - 1)) && (*cmdline != '\0'); ++argc) {
				argv[argc] = cmdline;
				while (*cmdline != ';' && *cmdline != '\0')
					++cmdline;
```
Remove `if (...) break;` below and `argc++` in line below for, should work fine, I think.

I mention this because of MISRA's `for loop should be well formed` rule.

Reply:
https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/24c96fa6795abe2aa12f7ecb43d4a2b89bc9b242/main.c#L72-L73
Above condition will not be true if there is a `;` or `\0`.

If `cmdline` points to `;`, it will change value to `\0` and starts pointing to the next byte. It is ok.
https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/24c96fa6795abe2aa12f7ecb43d4a2b89bc9b242/main.c#L75-L78

However, if `cmdline` points to `\0`, it should exit and not perform `line 78`. 
In Your proposition, `cmdline` will point to data outside a string what will cause that `for` loop will execute again, what is wrong.

Reply:
Ah, yes, you're right, not that simple. Ok, so let's drop this for now